File: src/commands/Tweet.ts
import { Command, createCommand } from 'commander';

import { TweetRepliesSortType } from '../enums/Tweet';
import { output } from '../helper/CliUtils';
import { TweetFilter } from '../models/args/FetchArgs';
import { Rettiwt } from '../Rettiwt';
import { ITweetFilter } from '../types/args/FetchArgs';

/**
 * Creates a new 'tweet' command which uses the given Rettiwt instance.
 *
 * @param rettiwt - The Rettiwt instance to use.
 * @returns The created 'tweet' command.
 */
function createTweetCommand(rettiwt: Rettiwt): Command {
	// Creating the 'tweet' command
	const tweet = createCommand('tweet').description('Access resources related to tweets');

	// Bookmark
	tweet
		.command('bookmark')
		.description('Bookmark a tweet')
		.argument('<id>', 'The tweet to bookmark')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.bookmark(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Details
	tweet
		.command('details')
		.description('Fetch the details of tweet/tweets with the given id/ids')
		.argument('<id>', 'The comma-separated list of IDs of tweets whose details are to be fetched')
		.action(async (id: string) => {
			try {
				// Getting the different IDs
				const ids: string[] = id.split(',');

				// If single ID given
				if (ids.length <= 1) {
					const details = await rettiwt.tweet.details(ids[0]);
					output(details);
				}
				// If multiple IDs give
				else {
					const details = await rettiwt.tweet.details(ids);
					output(details);
				}
			} catch (error) {
				output(error);
			}
		});

	// Like
	tweet
		.command('like')
		.description('Like a tweet')
		.argument('<id>', 'The tweet to like')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.like(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Likers
	tweet
		.command('likers')
		.description('Fetch the list of users who liked the given tweet. Only works for your own tweets')
		.argument('<id>', 'The id of the tweet')
		.argument('[count]', 'The number of likers to fetch')
		.argument('[cursor]', 'The cursor to the batch of likers to fetch')
		.action(async (id: string, count?: string, cursor?: string) => {
			try {
				const users = await rettiwt.tweet.likers(id, count ? parseInt(count) : undefined, cursor);
				output(users);
			} catch (error) {
				output(error);
			}
		});

	// Post
	tweet
		.command('post')
		.description('Post a tweet (text only)')
		.argument('<text>', 'The text to post as a tweet')
		.option('-m, --media [string]', 'Comma-separated list of ids of the media item(s) to be posted')
		.option('-q, --quote [string]', 'The id of the tweet to quote in the tweet to be posted')
		.option(
			'-r, --reply [string]',
			'The id of the tweet to which the reply is to be made, if the tweet is to be a reply',
		)
		.action(async (text: string, options?: { media?: string; quote?: string; reply?: string }) => {
			try {
				const result = await rettiwt.tweet.post({
					text: text,
					media: options?.media ? options?.media.split(',').map((item) => ({ id: item })) : undefined,
					quote: options?.quote,
					replyTo: options?.reply,
				});
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Replies
	tweet
		.command('replies')
		.description(
			'Fetch the list of replies to a tweet, with the first batch containing the whole thread, if the tweet is/part of a thread',
		)
		.argument('<id>', 'The id of the tweet')
		.argument('[cursor]', 'The cursor to the batch of replies to fetch')
		.option('-s, --sort-by <string>', 'Sort the tweets by likes, latest or relevance, default is latest')
		.action(async (id: string, cursor?: string, options?: { sortBy: string }) => {
			try {
				// Determining the sort type
				let sortType: TweetRepliesSortType | undefined = undefined;
				if (options?.sortBy === 'likes') {
					sortType = TweetRepliesSortType.LIKES;
				} else if (options?.sortBy === 'latest') {
					sortType = TweetRepliesSortType.LATEST;
				} else if (options?.sortBy === 'relevance') {
					sortType = TweetRepliesSortType.RELEVANCE;
				}

				const tweets = await rettiwt.tweet.replies(id, cursor, sortType);
				output(tweets);
			} catch (error) {
				output(error);
			}
		});

	// Retweet
	tweet
		.command('retweet')
		.description('Retweet a tweet')
		.argument('<id>', 'The tweet to retweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.retweet(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Retweeters
	tweet
		.command('retweeters')
		.description('Fetch the list of users who retweeted the given tweets')
		.argument('<id>', 'The id of the tweet')
		.argument('[count]', 'The number of retweeters to fetch')
		.argument('[cursor]', 'The cursor to the batch of retweeters to fetch')
		.action(async (id: string, count?: string, cursor?: string) => {
			try {
				const users = await rettiwt.tweet.retweeters(id, count ? parseInt(count) : undefined, cursor);
				output(users);
			} catch (error) {
				output(error);
			}
		});

	// Schedule
	tweet
		.command('schedule')
		.description('Schedule a tweet to be posted at a given date/time')
		.argument('<text>', 'The text to post as a tweet')
		.argument('<time>', 'The date/time at which the tweet is to be scheduled (valid date/time string)')
		.option('-m, --media [string]', 'Comma-separated list of ids of the media item(s) to be posted')
		.option('-q, --quote [string]', 'The id of the tweet to quote in the tweet to be posted')
		.option(
			'-r, --reply [string]',
			'The id of the tweet to which the reply is to be made, if the tweet is to be a reply',
		)
		.action(async (text: string, time: string, options?: { media?: string; quote?: string; reply?: string }) => {
			try {
				const result = await rettiwt.tweet.schedule({
					text: text,
					media: options?.media ? options?.media.split(',').map((item) => ({ id: item })) : undefined,
					quote: options?.quote,
					replyTo: options?.reply,
					scheduleFor: new Date(time),
				});
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Search
	tweet
		.command('search')
		.description('Fetch the list of tweets that match the given filter options')
		.argument('[count]', 'The number of tweets to fetch')
		.argument('[cursor]', 'The cursor to the batch of tweets to fetch')
		.option('-f, --from <string>', 'Matches the tweets made by the comma-separated list of given users')
		.option('-t, --to <string>', 'Matches the tweets made to the comma-separated list of given users')
		.option('-w, --words <string>', 'Matches the tweets containing the given comma-separated list of words')
		.option('-p, --phrase <string>', 'Matches the tweets containing the exact phrase')
		.option(
			'--optional-words <string>',
			'Matches the tweets containing any of the given comma-separated list of words',
		)
		.option(
			'--exclude-words <string>',
			'Matches the tweets that do not contain any of the give comma-separated list of words',
		)
		.option('-h, --hashtags <string>', 'Matches the tweets containing the given comma-separated list of hashtags')
		.option('--list <string>', 'Matches the tweets from the list with the given id')
		.option(
			'-m, --mentions <string>',
			'Matches the tweets that mention the given comma-separated list of usernames',
		)
		.option('-r, --min-replies <number>', 'Matches the tweets that have a minimum of given number of replies')
		.option('-l, --min-likes <number>', 'Matches the tweets that have a minimum of given number of likes')
		.option('-x, --min-retweets <number>', 'Matches the tweets that have a minimum of given number of retweets')
		.option('-q, --quoted <string>', 'Matches the tweets that quote the tweet with the given id')
		.option('--only-original', 'Matches tweets are original posts')
		.option('--only-replies', 'Matches tweets that are replies')
		.option('--only-text', 'Matches tweets that are only text')
		.option('--only-links', 'Matches tweets that only contain links like media, quotes, etc')
		.option('-s, --start <string>', 'Matches the tweets made since the given date (valid date/time string)')
		.option('-e, --end <string>', 'Matches the tweets made upto the given date (valid date/time string)')
		.option('--top', 'Matches top tweets instead of latest')
		.option('--stream', 'Stream the filtered tweets in pseudo-realtime')
		.option('-i, --interval <number>', 'The polling interval (in ms) to use for streaming. Default is 60000')
		.action(async (count?: string, cursor?: string, options?: TweetSearchOptions) => {
			try {
				// If search results are to be streamed
				if (options?.stream) {
					for await (const tweet of rettiwt.tweet.stream(
						new TweetSearchOptions(options).toTweetFilter(),
						options?.interval,
					)) {
						output(tweet);
					}
				}
				// If a normal search is to be done
				else {
					const tweets = await rettiwt.tweet.search(
						new TweetSearchOptions(options).toTweetFilter(),
						count ? parseInt(count) : undefined,
						cursor,
					);
					output(tweets);
				}
			} catch (error) {
				output(error);
			}
		});

	// Unbookmark
	tweet
		.command('unbookmark')
		.description('Unbookmark a tweet')
		.argument('<id>', 'The id of the tweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.unbookmark(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Unlike
	tweet
		.command('unlike')
		.description('Unlike a tweet')
		.argument('<id>', 'The id of the tweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.unlike(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Unpost
	tweet
		.command('unpost')
		.description('Unpost a tweet')
		.argument('<id>', 'The id of the tweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.unpost(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Unretweet
	tweet
		.command('unretweet')
		.description('Unretweet a tweet')
		.argument('<id>', 'The id of the tweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.unretweet(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Unschedule
	tweet
		.command('unschedule')
		.description('Unschedule a tweet')
		.argument('<id>', 'The id of the tweet')
		.action(async (id: string) => {
			try {
				const result = await rettiwt.tweet.unschedule(id);
				output(result);
			} catch (error) {
				output(error);
			}
		});

	// Upload
	tweet
		.command('upload')
		.description('Upload a media file and returns the alloted id (valid for 24 hrs)')
		.argument('<path>', 'The path to the media to upload')
		.action(async (path: string) => {
			try {
				const id = await rettiwt.tweet.upload(path);
				output(id);
			} catch (error) {
				output(error);
			}
		});

	return tweet;
}

/**
 * The search options supplied while searching for tweets.
 *
 * @remarks The search options are implementations of the ones offered by {@link TweetFilter}
 */
class TweetSearchOptions {
	public end?: string;
	public excludeWords?: string;
	public from?: string;
	public hashtags?: string;
	public interval?: number;
	public list?: string;
	public mentions?: string;
	public minLikes?: number;
	public minReplies?: number;
	public minRetweets?: number;
	public onlyLinks?: boolean = false;
	public onlyOriginal?: boolean = false;
	public onlyReplies?: boolean = false;
	public onlyText?: boolean = false;
	public optionalWords?: string;
	public phrase?: string;
	public quoted?: string;
	public start?: string;
	public stream?: boolean;
	public to?: string;
	public top?: boolean;
	public words?: string;

	/**
	 * Initializes a new object from the given options.
	 *
	 * @param options - The search options.
	 */
	public constructor(options?: TweetSearchOptions) {
		this.from = options?.from;
		this.to = options?.to;
		this.words = options?.words;
		this.phrase = options?.phrase;
		this.optionalWords = options?.optionalWords;
		this.excludeWords = options?.excludeWords;
		this.hashtags = options?.hashtags;
		this.list = options?.list;
		this.mentions = options?.mentions;
		this.minReplies = options?.minReplies;
		this.minLikes = options?.minLikes;
		this.minRetweets = options?.minRetweets;
		this.onlyLinks = options?.onlyLinks;
		this.onlyOriginal = options?.onlyOriginal;
		this.onlyReplies = options?.onlyReplies;
		this.onlyText = options?.onlyText;
		this.quoted = options?.quoted;
		this.start = options?.start;
		this.end = options?.end;
		this.stream = options?.stream;
		this.interval = options?.interval;
		this.top = options?.top;
	}

	/**
	 * Converts the filter options to a format recognizable by rettiwt-api.
	 *
	 * @returns The '{@link ITweetFilter}' representation of filter options.
	 */
	public toTweetFilter(): ITweetFilter {
		return new TweetFilter({
			fromUsers: this.from ? this.from.split(',') : undefined,
			toUsers: this.to ? this.to.split(',') : undefined,
			includeWords: this.words ? this.words.split(',') : undefined,
			includePhrase: this.phrase,
			optionalWords: this.optionalWords ? this.optionalWords.split(',') : undefined,
			excludeWords: this.excludeWords ? this.excludeWords.split(',') : undefined,
			hashtags: this.hashtags ? this.hashtags.split(',') : undefined,
			list: this.list,
			mentions: this.mentions ? this.mentions.split(',') : undefined,
			minReplies: this.minReplies,
			minLikes: this.minLikes,
			minRetweets: this.minRetweets,
			onlyLinks: this.onlyLinks,
			onlyOriginal: this.onlyOriginal,
			onlyReplies: this.onlyReplies,
			onlyText: this.onlyText,
			quoted: this.quoted,
			startDate: this.start ? new Date(this.start) : undefined,
			top: this.top,
			endDate: this.end ? new Date(this.end) : undefined,
		});
	}
}

export default createTweetCommand;

File: tweet.go
package main

import (
	"fmt"
	"github.com/spf13/cobra"
)

func createTweetCommand(rettiwt *Rettiwt) *cobra.Command {
	tweetCmd := &cobra.Command{
		Use:   "tweet",
		Short: "Access resources related to tweets",
	}

	detailsCmd := &cobra.Command{
		Use:   "details [id]",
		Short: "Fetch the details of tweet with the given id",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			details, err := rettiwt.Tweet.Details(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Details:", details)
		},
	}

	likeCmd := &cobra.Command{
		Use:   "like [id]",
		Short: "Like a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Like(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	listCmd := &cobra.Command{
		Use:   "list [id] [count] [cursor]",
		Short: "Fetch the list of tweets in the tweet list with the given id",
		Args:  cobra.MinimumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			var count int
			var cursor string
			if len(args) > 1 {
				count = args[1]
			}
			if len(args) > 2 {
				cursor = args[2]
			}
			tweets, err := rettiwt.Tweet.List(id, count, cursor)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Tweets:", tweets)
		},
	}

	postCmd := &cobra.Command{
		Use:   "post [text]",
		Short: "Post a tweet (text only)",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			text := args[0]
			media, _ := cmd.Flags().GetString("media")
			quote, _ := cmd.Flags().GetString("quote")
			reply, _ := cmd.Flags().GetString("reply")
			result, err := rettiwt.Tweet.Post(TweetArgs{
				Text:      text,
				Media:     media,
				Quote:     quote,
				ReplyTo:   reply,
			})
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}
	postCmd.Flags().String("media", "", "Comma-separated list of ids of the media item(s) to be posted")
	postCmd.Flags().String("quote", "", "The id of the tweet to quote in the tweet to be posted")
	postCmd.Flags().String("reply", "", "The id of the tweet to which the reply is to be made, if the tweet is to be a reply")

	retweetCmd := &cobra.Command{
		Use:   "retweet [id]",
		Short: "Retweet a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Retweet(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	retweetersCmd := &cobra.Command{
		Use:   "retweeters [id] [count] [cursor]",
		Short: "Fetch the list of users who retweeted the given tweets",
		Args:  cobra.MinimumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			var count int
			var cursor string
			if len(args) > 1 {
				count = args[1]
			}
			if len(args) > 2 {
				cursor = args[2]
			}
			retweeters, err := rettiwt.Tweet.Retweeters(id, count, cursor)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Retweeters:", retweeters)
		},
	}

	scheduleCmd := &cobra.Command{
		Use:   "schedule [text] [time]",
		Short: "Schedule a tweet to be posted at a given date/time",
		Args:  cobra.ExactArgs(2),
		Run: func(cmd *cobra.Command, args []string) {
			text := args[0]
			time := args[1]
			media, _ := cmd.Flags().GetString("media")
			quote, _ := cmd.Flags().GetString("quote")
			reply, _ := cmd.Flags().GetString("reply")
			result, err := rettiwt.Tweet.Schedule(TweetArgs{
				Text:        text,
				Media:       media,
				Quote:       quote,
				ReplyTo:     reply,
				ScheduleFor: time,
			})
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}
	scheduleCmd.Flags().String("media", "", "Comma-separated list of ids of the media item(s) to be posted")
	scheduleCmd.Flags().String("quote", "", "The id of the tweet to quote in the tweet to be posted")
	scheduleCmd.Flags().String("reply", "", "The id of the tweet to which the reply is to be made, if the tweet is to be a reply")

	searchCmd := &cobra.Command{
		Use:   "search [count] [cursor]",
		Short: "Fetch the list of tweets that match the given filter options",
		Args:  cobra.MinimumNArgs(0),
		Run: func(cmd *cobra.Command, args []string) {
			var count int
			var cursor string
			if len(args) > 0 {
				count = args[0]
			}
			if len(args) > 1 {
				cursor = args[1]
			}
			options := TweetSearchOptions{
				From:           cmd.Flag("from").Value.String(),
				To:             cmd.Flag("to").Value.String(),
				Words:          cmd.Flag("words").Value.String(),
				Phrase:         cmd.Flag("phrase").Value.String(),
				OptionalWords:  cmd.Flag("optional-words").Value.String(),
				ExcludeWords:   cmd.Flag("exclude-words").Value.String(),
				Hashtags:       cmd.Flag("hashtags").Value.String(),
				List:           cmd.Flag("list").Value.String(),
				Mentions:       cmd.Flag("mentions").Value.String(),
				MinReplies:     cmd.Flag("min-replies").Value.String(),
				MinLikes:       cmd.Flag("min-likes").Value.String(),
				MinRetweets:    cmd.Flag("min-retweets").Value.String(),
				Quoted:         cmd.Flag("quoted").Value.String(),
				ExcludeLinks:   cmd.Flag("exclude-links").Value.String(),
				ExcludeReplies: cmd.Flag("exclude-replies").Value.String(),
				Start:          cmd.Flag("start").Value.String(),
				End:            cmd.Flag("end").Value.String(),
				Top:            cmd.Flag("top").Value.String(),
				Stream:         cmd.Flag("stream").Value.String(),
				Interval:       cmd.Flag("interval").Value.String(),
			}
			if options.Stream {
				for tweet := range rettiwt.Tweet.Stream(options.ToTweetFilter(), options.Interval) {
					fmt.Println("Tweet:", tweet)
				}
			} else {
				tweets, err := rettiwt.Tweet.Search(options.ToTweetFilter(), count, cursor)
				if err != nil {
					fmt.Println("Error:", err)
					return
				}
				fmt.Println("Tweets:", tweets)
			}
		},
	}
	searchCmd.Flags().String("from", "", "Matches the tweets made by the comma-separated list of given users")
	searchCmd.Flags().String("to", "", "Matches the tweets made to the comma-separated list of given users")
	searchCmd.Flags().String("words", "", "Matches the tweets containing the given comma-separated list of words")
	searchCmd.Flags().String("phrase", "", "Matches the tweets containing the exact phrase")
	searchCmd.Flags().String("optional-words", "", "Matches the tweets containing any of the given comma-separated list of words")
	searchCmd.Flags().String("exclude-words", "", "Matches the tweets that do not contain any of the give comma-separated list of words")
	searchCmd.Flags().String("hashtags", "", "Matches the tweets containing the given comma-separated list of hashtags")
	searchCmd.Flags().String("list", "", "Matches the tweets from the list with the given id")
	searchCmd.Flags().String("mentions", "", "Matches the tweets that mention the given comma-separated list of usernames")
	searchCmd.Flags().String("min-replies", "", "Matches the tweets that have a minimum of given number of replies")
	searchCmd.Flags().String("min-likes", "", "Matches the tweets that have a minimum of given number of likes")
	searchCmd.Flags().String("min-retweets", "", "Matches the tweets that have a minimum of given number of retweets")
	searchCmd.Flags().String("quoted", "", "Matches the tweets that quote the tweet with the given id")
	searchCmd.Flags().String("exclude-links", "", "Matches tweets that do not contain links")
	searchCmd.Flags().String("exclude-replies", "", "Matches the tweets that are not replies")
	searchCmd.Flags().String("start", "", "Matches the tweets made since the given date (valid date/time string)")
	searchCmd.Flags().String("end", "", "Matches the tweets made upto the given date (valid date/time string)")
	searchCmd.Flags().String("top", "", "Matches top tweets instead of latest")
	searchCmd.Flags().String("stream", "", "Stream the filtered tweets in pseudo-realtime")
	searchCmd.Flags().String("interval", "", "The polling interval (in ms) to use for streaming. Default is 60000")

	unlikeCmd := &cobra.Command{
		Use:   "unlike [id]",
		Short: "Unlike a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Unlike(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	unpostCmd := &cobra.Command{
		Use:   "unpost [id]",
		Short: "Unpost a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Unpost(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	unretweetCmd := &cobra.Command{
		Use:   "unretweet [id]",
		Short: "Unretweet a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Unretweet(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	unscheduleCmd := &cobra.Command{
		Use:   "unschedule [id]",
		Short: "Unschedule a tweet",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			id := args[0]
			result, err := rettiwt.Tweet.Unschedule(id)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("Result:", result)
		},
	}

	uploadCmd := &cobra.Command{
		Use:   "upload [path]",
		Short: "Upload a media file and returns the alloted id (valid for 24 hrs)",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			path := args[0]
			id, err := rettiwt.Tweet.Upload(path)
			if err != nil {
				fmt.Println("Error:", err)
				return
			}
			fmt.Println("ID:", id)
		},
	}

	tweetCmd.AddCommand(detailsCmd)
	tweetCmd.AddCommand(likeCmd)
	tweetCmd.AddCommand(listCmd)
	tweetCmd.AddCommand(postCmd)
	tweetCmd.AddCommand(retweetCmd)
	tweetCmd.AddCommand(retweetersCmd)
	tweetCmd.AddCommand(scheduleCmd)
	tweetCmd.AddCommand(searchCmd)
	tweetCmd.AddCommand(unlikeCmd)
	tweetCmd.AddCommand(unpostCmd)
	tweetCmd.AddCommand(unretweetCmd)
	tweetCmd.AddCommand(unscheduleCmd)
	tweetCmd.AddCommand(uploadCmd)

	return tweetCmd
}

type TweetArgs struct {
	Text        string
	Media       string
	Quote       string
	ReplyTo     string
	ScheduleFor string
}

type TweetSearchOptions struct {
	From           string
	To             string
	Words          string
	Phrase         string
	OptionalWords  string
	ExcludeWords   string
	Hashtags       string
	List           string
	Mentions       string
	MinReplies     string
	MinLikes       string
	MinRetweets    string
	Quoted         string
	ExcludeLinks   string
	ExcludeReplies string
	Start          string
	End            string
	Top            string
	Stream         string
	Interval       string
}

func (options *TweetSearchOptions) ToTweetFilter() TweetFilter {
	return TweetFilter{
		FromUsers:      options.From,
		ToUsers:        options.To,
		IncludeWords:   options.Words,
		IncludePhrase:  options.Phrase,
		OptionalWords:  options.OptionalWords,
		ExcludeWords:   options.ExcludeWords,
		Hashtags:       options.Hashtags,
		List:           options.List,
		Mentions:       options.Mentions,
		MinReplies:     options.MinReplies,
		MinLikes:       options.MinLikes,
		MinRetweets:    options.MinRetweets,
		Quoted:         options.Quoted,
		Links:          options.ExcludeLinks,
		Replies:        options.ExcludeReplies,
		StartDate:      options.Start,
		EndDate:        options.End,
		Top:            options.Top,
	}
}
